src/components/name-tag.js

```diff
@@ -0,0 +1,377 @@
+    this.nametagElPosY = this.nametagHeight;
+    this.size = this.nametagTextEl.components["text"].getSize();
+    this.size.x = Math.max(this.size.x, NAMETAG_MIN_WIDTH);
+    this.updateNameTag();
+  },
+
+  onAnalyserVolumeUpdated({ detail: { volume } }) {
+    let isTalking = false;
+    this.volume = volume;
+    this.volumeAvg.push(Date.now(), volume);
+    if (!this.playerInfo.data.muted) {
+      const average = this.volumeAvg.movingAverage();
+      isTalking = average > 0.01;
+    }
+    this.wasTalking = this.isTalking;
+    this.isTalking = isTalking;
+    if (this.nametagVisibility === "showSpeaking") {
+      if (!this.isTalking && this.wasTalking) {
+        this.frozenTimer = setTimeout(() => {
+          this.wasNametagVisible = this.isNametagVisible;
+          this.isNametagVisible = false;
+          this.isNametagVisible !== this.wasNametagVisible && this.updateNameTag();
+        }, 1000);
+      } else if (this.isTalking && !this.wasTalking) {
+        clearTimeout(this.frozenTimer);
+        this.wasNametagVisible = this.isNametagVisible;
+        this.isNametagVisible = true;
+        this.isNametagVisible !== this.wasNametagVisible && this.updateNameTag();
+      }
+    }
+    this.isNametagVisible && this.isTalking !== this.wasTalking && this.updateBorder();
+    this.isNametagVisible && this.updateVolume();
+  },
+
+  onStateChanged() {
+    this.updateNameTagVisibility();
+  },
+
+  updateNameTagVisibility() {
+    this.nametagVisibilityDistance =
+      this.store.state.preferences.nametagVisibilityDistance !== undefined
+        ? this.store.state.preferences.nametagVisibilityDistance
+        : NAMETAG_VISIBILITY_DISTANCE_DEFAULT;
+    this.wasNametagVisible = this.isNametagVisible;
+    let prefNametagVisibility = this.store.state.preferences.nametagVisibility === undefined;
+    if (this.nametagVisibility === "showNone") {
+      prefNametagVisibility = false;
+    } else if (this.nametagVisibility === "showAll") {
+      prefNametagVisibility = true;
+    } else if (this.nametagVisibility === "showFrozen") {
+      prefNametagVisibility = this.el.sceneEl.is("frozen");
+    } else if (this.nametagVisibility === "showSpeaking") {
+      prefNametagVisibility = this.isTalking;
+    } else {
+      prefNametagVisibility = true;
+    }
+    this.isNametagVisible = prefNametagVisibility && this.tmpNametagVisible;
+    this.updateNameTag();
+  },
+
+  updateNameTag() {
+    if (!this.model) return;
+    this.updateDisplayName();
+    this.updateContainer();
+    this.updateVolume();
+    this.updateState();
+  },
+
+  updateContainer() {
+    this.nametagBackgroundEl.setAttribute("visible", this.isNametagVisible);
+    this.nametagBackgroundEl.setAttribute("slice9", {
+      width: this.size.x + NAMETAG_BACKGROUND_PADDING * 2,
+      height: NAMETAG_HEIGHT
+    });
+    this.updateBorder();
+  },
+
+  updateVolume() {
+    this.nametagVolumeEl.setAttribute("visible", this.isTalking && this.isNametagVisible);
```

Comment:
As elsewhere, skip the aframe stuff, but also this object's visibility should only depend on `this.isTalking` (or better yet directly on the analyzer `isTalking` state). The nametag visibility should be handled by hiding the root.

Reply:
Updated